From f692f65e8a5d523001d86684004d4210949bc701 Mon Sep 17 00:00:00 2001
From: Oliver Brown <oliver.brown@freescale.com>
Date: Thu, 29 Nov 2012 15:21:26 -0600
Subject: [PATCH 04/50] ENGR00235350 - Quadrant CPU h264 test failed on MX53
 smd Android JellyBean

Commit on behalf of Chingling Wang:
OMX Buffers must be returned on a flush.

Signed-off-by: Oliver Brown <oliver.brown@freescale.com>
---
 media/libstagefright/Android.mk   |  4 ++++
 media/libstagefright/OMXCodec.cpp | 13 +++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/media/libstagefright/Android.mk b/media/libstagefright/Android.mk
index 6a2a696..5bd5c20 100644
--- a/media/libstagefright/Android.mk
+++ b/media/libstagefright/Android.mk
@@ -107,6 +107,10 @@ LOCAL_SRC_FILES += \
         chromium_http_stub.cpp
 LOCAL_CPPFLAGS += -DCHROMIUM_AVAILABLE=1
 
+ifeq ($(BOARD_SOC_CLASS), IMX5X)
+   LOCAL_CFLAGS += -DIMX5X
+endif
+
 LOCAL_SHARED_LIBRARIES += libstlport
 include external/stlport/libstlport.mk
 
diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index dcaef76..646c93a 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2009 The Android Open Source Project
+ * Copyright (C) 2012 Freescale Semiconductor, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -2606,6 +2607,18 @@ void OMXCodec::onCmdComplete(OMX_COMMANDTYPE cmd, OMX_U32 data) {
 
                     drainInputBuffers();
                     fillOutputBuffers();
+                    #if defined (IMX5X)
+                    Vector<BufferInfo> *buffers = &mPortBuffers[kPortIndexOutput];
+                     for (size_t i = 0; i < buffers->size(); ++i) {
+                          BufferInfo *info = &buffers->editItemAt(i);
+                          if(info->mStatus == OWNED_BY_NATIVE_WINDOW){
+                               // Give the buffer to the OMX node to fill. MX53 vpu firmware needs all buffer returned when flush
+                                 info->mStatus = OWNED_BY_US;
+                                 fillOutputBuffer(info);
+                                 info->mStatus = OWNED_BY_NATIVE_WINDOW;
+                          }
+                        }
+                    #endif
                 }
 
                 if (mOutputPortSettingsChangedPending) {
-- 
1.8.0

